File: app/src/main/java/com/privacylion/btcdid/BreezWalletManager.kt
package com.privacylion.btcdid

import android.content.Context
import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import breez_sdk_spark.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.io.File
import java.security.KeyStore
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Breez SDK Spark Wallet Manager for SignedByMe
 * 
 * Handles Lightning wallet operations using the Breez SDK Nodeless (Spark) implementation:
 * - Wallet initialization and mnemonic management
 * - BOLT11 invoice generation for receiving payments
 * - Payment status monitoring
 * - Balance tracking
 */
class BreezWalletManager(private val context: Context) {

    companion object {
        private const val TAG = "BreezWalletManager"
        private const val KEYSTORE_ALIAS = "breez_mnemonic_wrap_v1"
        private const val MNEMONIC_FILE = "breez_mnemonic_wrapped.bin"
        private const val ANDROID_KEYSTORE = "AndroidKeyStore"
        
        // Breez API key - get yours from https://breez.technology/request-api-key/
        private const val BREEZ_API_KEY = "" // TODO: Add your Breez API key
    }

    private val rng = SecureRandom()
    private var sdk: BreezSdk? = null
    
    // State flows for UI observation
    private val _walletState = MutableStateFlow<WalletState>(WalletState.Uninitialized)
    val walletState: StateFlow<WalletState> = _walletState
    
    private val _balanceSats = MutableStateFlow(0L)
    val balanceSats: StateFlow<Long> = _balanceSats

    sealed class WalletState {
        object Uninitialized : WalletState()
        object Initializing : WalletState()
        data class Ready(val nodeId: String) : WalletState()
        data class Error(val message: String) : WalletState()
    }

    /**
     * Data class for invoice creation result
     */
    data class InvoiceResult(
        val bolt11: String,
        val paymentHash: String,
        val amountSats: Long,
        val description: String,
        val feeSats: Long
    )

    /**
     * Initialize or restore the wallet
     * If no mnemonic exists, creates a new one
     */
    suspend fun initializeWallet(): Result<String> = withContext(Dispatchers.IO) {
        try {
            _walletState.value = WalletState.Initializing
            
            // Check API key
            if (BREEZ_API_KEY.isBlank()) {
                _walletState.value = WalletState.Error("Breez API key not configured")
                return@withContext Result.failure(IllegalStateException("Breez API key not configured. Get one from https://breez.technology/request-api-key/"))
            }
            
            // Ensure keystore key exists
            ensureKeystoreKey()
            
            // Get or create mnemonic
            val mnemonic = getOrCreateMnemonic()
            
            // Create seed from mnemonic
            val seed = Seed.Mnemonic(mnemonic, null)
            
            // Set up storage directory
            val storageDir = File(context.filesDir, "breez_spark").apply { mkdirs() }
            
            // Create config
            val config = defaultConfig(Network.MAINNET)
            config.apiKey = BREEZ_API_KEY
            
            // Connect to Breez SDK
            val connectRequest = ConnectRequest(
                config = config,
                seed = seed,
                storageDir = storageDir.absolutePath
            )
            
            sdk = connect(connectRequest)
            
            // Get wallet info
            val info = sdk?.getInfo()
            val pubkey = info?.pubkey ?: "unknown"
            
            // Update balance
            val balanceMsat = info?.balanceMsat ?: 0UL
            _balanceSats.value = (balanceMsat / 1000UL).toLong()
            
            _walletState.value = WalletState.Ready(pubkey)
            
            // Start listening for events
            startEventListener()
            
            Result.success(pubkey)
        } catch (e: Exception) {
            val errorMsg = e.message ?: "Unknown error initializing wallet"
            _walletState.value = WalletState.Error(errorMsg)
            Result.failure(e)
        }
    }

    /**
     * Check if wallet is already set up (mnemonic exists)
     */
    fun isWalletSetUp(): Boolean {
        return context.getFileStreamPath(MNEMONIC_FILE)?.exists() == true
    }

    /**
     * Get wallet balance in sats
     */
    suspend fun getBalance(): Long = withContext(Dispatchers.IO) {
        try {
            val info = sdk?.getInfo()
            val balanceMsat = info?.balanceMsat ?: 0UL
            val balance = (balanceMsat / 1000UL).toLong()
            _balanceSats.value = balance
            balance
        } catch (e: Exception) {
            0L
        }
    }

    /**
     * Generate a BOLT11 Lightning invoice for receiving payment
     */
    suspend fun createInvoice(
        amountSats: Long,
        description: String
    ): Result<InvoiceResult> = withContext(Dispatchers.IO) {
        try {
            val sdkInstance = sdk ?: return@withContext Result.failure(
                IllegalStateException("Wallet not initialized")
            )
            
            // Create BOLT11 invoice request
            val paymentMethod = ReceivePaymentMethod.Bolt11Invoice(
                description = description,
                amountSats = amountSats.toULong(),
                expirySecs = 3600u // 1 hour expiry
            )
            
            val request = ReceivePaymentRequest(paymentMethod)
            val response = sdkInstance.receivePayment(request)
            
            // Extract payment hash from the invoice
            // The paymentRequest is the BOLT11 string
            val bolt11 = response.paymentRequest
            val feeSats = response.fee?.toLong() ?: 0L
            
            // Parse payment hash from BOLT11 (it's embedded in the invoice)
            // For now, we'll use a hash of the bolt11 as identifier
            // The SDK should provide this, but we'll derive it
            val paymentHash = derivePaymentHash(bolt11)
            
            Result.success(
                InvoiceResult(
                    bolt11 = bolt11,
                    paymentHash = paymentHash,
                    amountSats = amountSats,
                    description = description,
                    feeSats = feeSats
                )
            )
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Get a Bitcoin address for on-chain receiving
     */
    suspend fun getBitcoinAddress(): Result<String> = withContext(Dispatchers.IO) {
        try {
            val sdkInstance = sdk ?: return@withContext Result.failure(
                IllegalStateException("Wallet not initialized")
            )
            
            val request = ReceivePaymentRequest(ReceivePaymentMethod.BitcoinAddress)
            val response = sdkInstance.receivePayment(request)
            
            Result.success(response.paymentRequest)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * List recent payments
     */
    suspend fun listPayments(limit: Int = 10): List<Payment> = withContext(Dispatchers.IO) {
        try {
            val sdkInstance = sdk ?: return@withContext emptyList()
            val request = ListPaymentsRequest(
                limit = limit.toUInt(),
                offset = null,
                filters = null
            )
            sdkInstance.listPayments(request)
        } catch (e: Exception) {
            emptyList()
        }
    }

    /**
     * Get the mnemonic for backup purposes
     * Only call this when user explicitly requests backup
     */
    fun getMnemonicForBackup(): String? {
        return try {
            val wrapped = context.openFileInput(MNEMONIC_FILE).use { it.readBytes() }
            val mnemonicBytes = unwrapData(wrapped)
            String(mnemonicBytes, Charsets.UTF_8)
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Disconnect and cleanup
     */
    suspend fun disconnect() = withContext(Dispatchers.IO) {
        try {
            sdk?.disconnect()
            sdk = null
            _walletState.value = WalletState.Uninitialized
        } catch (e: Exception) {
            // Ignore disconnect errors
        }
    }

    // ==================== Private Methods ====================

    private fun getOrCreateMnemonic(): String {
        val mnemonicFile = context.getFileStreamPath(MNEMONIC_FILE)
        
        return if (mnemonicFile?.exists() == true) {
            // Load existing mnemonic
            val wrapped = context.openFileInput(MNEMONIC_FILE).use { it.readBytes() }
            val mnemonicBytes = unwrapData(wrapped)
            String(mnemonicBytes, Charsets.UTF_8)
        } else {
            // Generate new 12-word mnemonic
            val mnemonic = generateMnemonic()
            
            // Wrap and save
            val mnemonicBytes = mnemonic.toByteArray(Charsets.UTF_8)
            val wrapped = wrapData(mnemonicBytes)
            context.openFileOutput(MNEMONIC_FILE, Context.MODE_PRIVATE).use { 
                it.write(wrapped) 
            }
            
            // Zeroize local copy
            java.util.Arrays.fill(mnemonicBytes, 0)
            
            mnemonic
        }
    }

    /**
     * Generate a 12-word BIP39 mnemonic
     * Uses SecureRandom for entropy
     */
    private fun generateMnemonic(): String {
        // BIP39 wordlist (English) - we'll use the SDK's built-in generation if available
        // For now, generate 128 bits of entropy for 12 words
        val entropy = ByteArray(16)
        rng.nextBytes(entropy)
        
        // Convert to mnemonic using BIP39
        // The Breez SDK should handle this, but we need raw words
        // For initial implementation, we'll use a simple approach
        // In production, use a proper BIP39 library
        return entropyToMnemonic(entropy)
    }

    /**
     * Convert entropy bytes to BIP39 mnemonic words
     * This is a simplified implementation - in production use a proper BIP39 library
     */
    private fun entropyToMnemonic(entropy: ByteArray): String {
        // BIP39 English wordlist (first 100 words for demo - full list has 2048)
        // In production, include the full BIP39 wordlist
        val wordlist = listOf(
            "abandon", "ability", "able", "about", "above", "absent", "absorb", "abstract",
            "absurd", "abuse", "access", "accident", "account", "accuse", "achieve", "acid",
            "acoustic", "acquire", "across", "act", "action", "actor", "actress", "actual",
            "adapt", "add", "addict", "address", "adjust", "admit", "adult", "advance",
            "advice", "aerobic", "affair", "afford", "afraid", "again", "age", "agent",
            "agree", "ahead", "aim", "air", "airport", "aisle", "alarm", "album",
            "alcohol", "alert", "alien", "all", "alley", "allow", "almost", "alone",
            "alpha", "already", "also", "alter", "always", "amateur", "amazing", "among",
            "amount", "amused", "analyst", "anchor", "ancient", "anger", "angle", "angry",
            "animal", "ankle", "announce", "annual", "another", "answer", "antenna", "antique",
            "anxiety", "any", "apart", "apology", "appear", "apple", "approve", "april",
            "arch", "arctic", "area", "arena", "argue", "arm", "armed", "armor",
            "army", "around", "arrange", "arrest", "arrive", "arrow", "art", "artefact",
            "artist", "artwork", "ask", "aspect", "assault", "asset", "assist", "assume",
            "asthma", "athlete", "atom", "attack", "attend", "attitude", "attract", "auction",
            "audit", "august", "aunt", "author", "auto", "autumn", "average", "avocado",
            "avoid", "awake", "aware", "away", "awesome", "awful", "awkward", "axis",
            "baby", "bachelor", "bacon", "badge", "bag", "balance", "balcony", "ball",
            "bamboo", "banana", "banner", "bar", "barely", "bargain", "barrel", "base",
            "basic", "basket", "battle", "beach", "bean", "beauty", "because", "become",
            "beef", "before", "begin", "behave", "behind", "believe", "below", "belt",
            "bench", "benefit", "best", "betray", "better", "between", "beyond", "bicycle",
            "bid", "bike", "bind", "biology", "bird", "birth", "bitter", "black",
            "blade", "blame", "blanket", "blast", "bleak", "bless", "blind", "blood",
            "blossom", "blouse", "blue", "blur", "blush", "board", "boat", "body",
            "boil", "bomb", "bone", "bonus", "book", "boost", "border", "boring",
            "borrow", "boss", "bottom", "bounce", "box", "boy", "bracket", "brain",
            "brand", "brass", "brave", "bread", "breeze", "brick", "bridge", "brief",
            "bright", "bring", "brisk", "broccoli", "broken", "bronze", "broom", "brother",
            "brown", "brush", "bubble", "buddy", "budget", "buffalo", "build", "bulb",
            "bulk", "bullet", "bundle", "bunker", "burden", "burger", "burst", "bus",
            "business", "busy", "butter", "buyer", "buzz", "cabbage", "cabin", "cable",
            "cactus", "cage", "cake", "call", "calm", "camera", "camp", "can",
            "canal", "cancel", "candy", "cannon", "canoe", "canvas", "canyon", "capable",
            "capital", "captain", "car", "carbon", "card", "cargo", "carpet", "carry",
            "cart", "case", "cash", "casino", "castle", "casual", "cat", "catalog",
            "catch", "category", "cattle", "caught", "cause", "caution", "cave", "ceiling"
        )
        
        // Generate 12 words from entropy
        val words = mutableListOf<String>()
        val md = java.security.MessageDigest.getInstance("SHA-256")
        val hash = md.digest(entropy)
        
        // Use hash bytes to select words (simplified)
        for (i in 0 until 12) {
            val index = ((hash[i].toInt() and 0xFF) + (hash[(i + 1) % hash.size].toInt() and 0xFF)) % wordlist.size
            words.add(wordlist[index])
        }
        
        return words.joinToString(" ")
    }

    /**
     * Derive a payment hash identifier from BOLT11 invoice
     */
    private fun derivePaymentHash(bolt11: String): String {
        val md = java.security.MessageDigest.getInstance("SHA-256")
        val hash = md.digest(bolt11.toByteArray(Charsets.UTF_8))
        return hash.joinToString("") { "%02x".format(it) }
    }

    private fun startEventListener() {
        // The Spark SDK handles events internally
        // We can add a listener if needed for real-time updates
        CoroutineScope(Dispatchers.IO).launch {
            while (sdk != null) {
                delay(30_000) // Check every 30 seconds
                try {
                    getBalance()
                } catch (e: Exception) {
                    // Ignore errors during background refresh
                }
            }
        }
    }

    private fun ensureKeystoreKey() {
        try {
            val ks = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }
            if (ks.containsAlias(KEYSTORE_ALIAS)) return

            val specBuilder = KeyGenParameterSpec.Builder(
                KEYSTORE_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setUserAuthenticationRequired(false) // TODO: Enable with BiometricPrompt
                .setRandomizedEncryptionRequired(true)

            if (Build.VERSION.SDK_INT >= 28) {
                try { specBuilder.setUnlockedDeviceRequired(true) } catch (_: Throwable) {}
                try { specBuilder.setIsStrongBoxBacked(true) } catch (_: Throwable) {}
            }

            val kg = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE)
            kg.init(specBuilder.build())
            kg.generateKey()
        } catch (_: Throwable) {
            // Fallback handled in getAesKey()
        }
    }

    private fun getAesKey(): SecretKey {
        return try {
            val ks = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }
            (ks.getKey(KEYSTORE_ALIAS, null) as SecretKey?) ?: createFallbackKey()
        } catch (_: Throwable) {
            createFallbackKey()
        }
    }

    private fun createFallbackKey(): SecretKey {
        val fallbackFile = "breez_aes_fallback.bin"
        val file = context.getFileStreamPath(fallbackFile)
        
        return if (file?.exists() == true) {
            val b = context.openFileInput(fallbackFile).use { it.readBytes() }
            SecretKeySpec(b, "AES")
        } else {
            val b = ByteArray(32).also { rng.nextBytes(it) }
            context.openFileOutput(fallbackFile, Context.MODE_PRIVATE).use { it.write(b) }
            SecretKeySpec(b, "AES")
        }
    }

    private fun wrapData(plain: ByteArray): ByteArray {
        val secret = getAesKey()
        val iv = ByteArray(12).also { rng.nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, secret, GCMParameterSpec(128, iv))
        val ct = cipher.doFinal(plain)
        return iv + ct
    }

    private fun unwrapData(wrapped: ByteArray): ByteArray {
        require(wrapped.size > 12) { "wrapped data too short" }
        val secret = getAesKey()
        val iv = wrapped.copyOfRange(0, 12)
        val ct = wrapped.copyOfRange(12, wrapped.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, secret, GCMParameterSpec(128, iv))
        return cipher.doFinal(ct)
    }
}

File: DID_BTC/DID_BTC/Managers/BreezWalletManager.swift
// BreezWalletManager.swift - Lightning Wallet Integration
// SignedByMe iOS - Mirrors Android BreezWalletManager.kt
// Uses Breez SDK Spark for Lightning payments

import Foundation
import Combine

/// Wallet connection state
enum WalletState: Equatable {
    case disconnected
    case connecting
    case connected(balance: Int64)
    case error(String)
    
    var isConnected: Bool {
        if case .connected = self { return true }
        return false
    }
}

/// Manages Breez SDK Spark wallet for Lightning payments
@MainActor
final class BreezWalletManager: ObservableObject {
    
    // MARK: - Published Properties
    
    @Published private(set) var state: WalletState = .disconnected
    @Published private(set) var balanceSats: Int64 = 0
    @Published private(set) var sparkAddress: String?
    @Published private(set) var transactions: [WalletTransaction] = []
    @Published private(set) var btcPriceUsd: Double = 0
    
    // MARK: - Private Properties
    
    private let keychain = KeychainManager.shared
    private var isInitialized = false
    
    // Note: In production, import BreezSDK and use real SDK calls
    // For now, we define the interface for when SDK is integrated
    
    // MARK: - Initialization
    
    init() {
        Task {
            await loadExistingWallet()
        }
    }
    
    // MARK: - Wallet Setup
    
    /// Create a new wallet with BIP39 mnemonic
    func createWallet() async throws -> [String] {
        state = .connecting
        
        do {
            // Generate 12-word BIP39 mnemonic
            let mnemonic = generateMnemonic()
            
            // Store mnemonic securely
            try keychain.save(mnemonic.joined(separator: " "), for: .walletMnemonic)
            
            // Initialize SDK with mnemonic
            try await initializeSDK(mnemonic: mnemonic)
            
            print("✅ Created new wallet")
            return mnemonic
        } catch {
            state = .error(error.localizedDescription)
            throw error
        }
    }
    
    /// Restore wallet from mnemonic
    func restoreWallet(mnemonic: [String]) async throws {
        state = .connecting
        
        guard mnemonic.count == 12 || mnemonic.count == 24 else {
            throw WalletError.invalidMnemonic
        }
        
        do {
            // Store mnemonic
            try keychain.save(mnemonic.joined(separator: " "), for: .walletMnemonic)
            
            // Initialize SDK
            try await initializeSDK(mnemonic: mnemonic)
            
            print("✅ Restored wallet from mnemonic")
        } catch {
            state = .error(error.localizedDescription)
            throw error
        }
    }
    
    /// Load existing wallet on app start
    private func loadExistingWallet() async {
        guard let mnemonicString = try? keychain.loadString(.walletMnemonic) else {
            print("📱 No existing wallet found")
            return
        }
        
        let mnemonic = mnemonicString.split(separator: " ").map(String.init)
        
        do {
            state = .connecting
            try await initializeSDK(mnemonic: mnemonic)
            print("📱 Loaded existing wallet")
        } catch {
            state = .error(error.localizedDescription)
            print("❌ Failed to load wallet: \(error)")
        }
    }
    
    /// Initialize Breez SDK
    private func initializeSDK(mnemonic: [String]) async throws {
        // TODO: Replace with real Breez SDK initialization
        // Example Breez SDK Swift code:
        // let config = BreezConfig(...)
        // let sdk = try await BreezSDK.connect(config: config, mnemonic: mnemonic.joined(separator: " "))
        
        // For now, simulate initialization
        try await Task.sleep(nanoseconds: 500_000_000)
        
        // Mock values - replace with real SDK calls
        balanceSats = 0
        sparkAddress = "sp1q\(UUID().uuidString.prefix(32).lowercased())"
        state = .connected(balance: balanceSats)
        isInitialized = true
        
        // Fetch BTC price
        await fetchBtcPrice()
    }
    
    // MARK: - Invoice Management
    
    /// Create a Lightning invoice for receiving
    func createInvoice(amountSats: Int64, description: String) async throws -> String {
        guard isInitialized else {
            throw WalletError.notInitialized
        }
        
        // TODO: Replace with real Breez SDK call
        // let invoice = try await sdk.receivePayment(amountSats: amountSats, description: description)
        // return invoice.bolt11
        
        // Mock invoice for development
        let mockInvoice = "lnbc\(amountSats)n1p\(UUID().uuidString.prefix(50).lowercased())"
        print("📝 Created invoice for \(amountSats) sats")
        return mockInvoice
    }
    
    /// Parse and pay a BOLT11 invoice
    func payInvoice(_ bolt11: String) async throws -> String {
        guard isInitialized else {
            throw WalletError.notInitialized
        }
        
        // TODO: Replace with real Breez SDK call
        // let payment = try await sdk.sendPayment(bolt11: bolt11)
        // return payment.paymentPreimage
        
        // Mock payment for development
        try await Task.sleep(nanoseconds: 1_000_000_000)
        let mockPreimage = "mock_preimage_\(UUID().uuidString)"
        print("💸 Paid invoice")
        return mockPreimage
    }
    
    /// Check if a payment has been received (by payment hash)
    func isPaymentReceived(paymentHash: String) async -> Bool {
        guard isInitialized else { return false }
        
        // TODO: Replace with real Breez SDK call
        // return await sdk.isPaymentReceived(paymentHash: paymentHash)
        
        // Mock check
        return false
    }
    
    // MARK: - Balance & Transactions
    
    /// Refresh wallet balance
    func refreshBalance() async throws {
        guard isInitialized else {
            throw WalletError.notInitialized
        }
        
        // TODO: Replace with real Breez SDK call
        // let nodeInfo = try await sdk.nodeInfo()
        // balanceSats = nodeInfo.channelsBalanceSats + nodeInfo.onchainBalanceSats
        
        state = .connected(balance: balanceSats)
    }
    
    /// Fetch transaction history
    func fetchTransactions() async throws {
        guard isInitialized else {
            throw WalletError.notInitialized
        }
        
        // TODO: Replace with real Breez SDK call
        // let payments = try await sdk.listPayments()
        // transactions = payments.map { ... }
        
        // Mock transactions for development
        transactions = []
    }
    
    // MARK: - Mnemonic
    
    /// Get mnemonic for backup display (requires biometric)
    func getMnemonic() async throws -> [String] {
        // Require biometric auth
        let authenticated = try await keychain.authenticateWithBiometrics(
            reason: "Authenticate to view recovery phrase"
        )
        guard authenticated else {
            throw WalletError.authenticationFailed
        }
        
        let mnemonicString = try keychain.loadString(.walletMnemonic)
        return mnemonicString.split(separator: " ").map(String.init)
    }
    
    /// Delete wallet (for testing/reset)
    func deleteWallet() throws {
        try keychain.delete(.walletMnemonic)
        try keychain.delete(.walletSeed)
        
        state = .disconnected
        balanceSats = 0
        sparkAddress = nil
        transactions = []
        isInitialized = false
    }
    
    // MARK: - BTC Price
    
    /// Fetch current BTC price
    func fetchBtcPrice() async {
        do {
            btcPriceUsd = try await APIService.shared.getBtcPrice()
        } catch {
            print("⚠️ Failed to fetch BTC price: \(error)")
        }
    }
    
    /// Convert sats to USD
    func satsToUsd(_ sats: Int64) -> Double {
        guard btcPriceUsd > 0 else { return 0 }
        return Double(sats) / 100_000_000 * btcPriceUsd
    }
    
    /// Format sats for display
    func formatSats(_ sats: Int64) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        return formatter.string(from: NSNumber(value: sats)) ?? "\(sats)"
    }
    
    /// Format USD for display
    func formatUsd(_ amount: Double) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: NSNumber(value: amount)) ?? "$\(amount)"
    }
    
    // MARK: - Private Helpers
    
    /// Generate BIP39 mnemonic
    private func generateMnemonic() -> [String] {
        // BIP39 word list (abbreviated - full list has 2048 words)
        // TODO: Use proper BIP39 library
        let wordList = [
            "abandon", "ability", "able", "about", "above", "absent", "absorb", "abstract",
            "absurd", "abuse", "access", "accident", "account", "accuse", "achieve", "acid",
            "acoustic", "acquire", "across", "act", "action", "actor", "actress", "actual",
            "adapt", "add", "addict", "address", "adjust", "admit", "adult", "advance",
            "advice", "aerobic", "affair", "afford", "afraid", "again", "age", "agent",
            "agree", "ahead", "aim", "air", "airport", "aisle", "alarm", "album",
            "alcohol", "alert", "alien", "all", "alley", "allow", "almost", "alone",
            "alpha", "already", "also", "alter", "always", "amateur", "amazing", "among"
        ]
        
        // Generate random 12 words
        var mnemonic: [String] = []
        for _ in 0..<12 {
            let randomIndex = Int.random(in: 0..<wordList.count)
            mnemonic.append(wordList[randomIndex])
        }
        
        return mnemonic
    }
}

// MARK: - Wallet Errors

enum WalletError: Error, LocalizedError {
    case notInitialized
    case invalidMnemonic
    case authenticationFailed
    case invoiceFailed
    case paymentFailed
    case insufficientBalance
    
    var errorDescription: String? {
        switch self {
        case .notInitialized: return "Wallet not initialized"
        case .invalidMnemonic: return "Invalid recovery phrase"
        case .authenticationFailed: return "Authentication failed"
        case .invoiceFailed: return "Failed to create invoice"
        case .paymentFailed: return "Payment failed"
        case .insufficientBalance: return "Insufficient balance"
        }
    }
}

// MARK: - Computed Properties

extension BreezWalletManager {
    
    /// Check if wallet is set up
    var hasWallet: Bool {
        keychain.hasWallet
    }
    
    /// Balance formatted for display
    var formattedBalance: String {
        formatSats(balanceSats)
    }
    
    /// Balance in USD formatted
    var formattedUsdBalance: String {
        formatUsd(satsToUsd(balanceSats))
    }
}
